src/NUnitFramework/framework/Attributes/TestCaseAttribute.cs

```diff
@@ -360,6 +301,79 @@ private TestCaseParameters GetParametersForTestCase(IMethodInfo method)
+           int argsProvided = parms.Arguments.Length;
+
+            // Special handling for params arguments
+            if (argsNeeded > 0 && argsProvided >= argsNeeded - 1)
+            {
+                IParameterInfo lastParameter = parameters[argsNeeded - 1];
+                Type lastParameterType = lastParameter.ParameterType;
+                Type elementType = lastParameterType.GetElementType();
+
+                if (lastParameterType.IsArray && lastParameter.IsDefined<ParamArrayAttribute>(false))
+                {
+                    if (argsProvided == argsNeeded)
+                    {
+                        Type lastArgumentType = parms.Arguments[argsProvided - 1].GetType();
+                        if (!lastParameterType.GetTypeInfo().IsAssignableFrom(lastArgumentType.GetTypeInfo()))
+                        {
+                            Array array = Array.CreateInstance(elementType, 1);
+                            array.SetValue(parms.Arguments[argsProvided - 1], 0);
+                            parms.Arguments[argsProvided - 1] = array;
+                        }
+                    }
+                    else
+                    {
+                        object[] newArglist = new object[argsNeeded];
+                        for (int i = 0; i < argsNeeded && i < argsProvided; i++)
+                            newArglist[i] = parms.Arguments[i];
+
+                        int length = argsProvided - argsNeeded + 1;
+                        Array array = Array.CreateInstance(elementType, length);
+                        for (int i = 0; i < length; i++)
+                            array.SetValue(parms.Arguments[argsNeeded + i - 1], i);
+
+                        newArglist[argsNeeded - 1] = array;
+                        parms.Arguments = newArglist;
+                        argsProvided = argsNeeded;
+                    }
+                }
+            }
+
+#if !NETCF
+            //Special handling for optional parameters
+            if (argsProvided < argsNeeded)
+            {
+                object[] newArgList = new object[parameters.Length];
+                Array.Copy(parms.Arguments, newArgList, argsProvided);
+
+                //Fill with Type.Missing for remaining required parameters where optional
+                for (var i = parms.Arguments.Length; i < parameters.Length; i++)
+                {
+                    if (parameters[i].IsOptional)
+                        newArgList[i] = Type.Missing;
+                    else
+                    {
+                        if (i < parms.Arguments.Length)
+                            newArgList[i] = parms.Arguments[i];
+                        else
+                            throw new TargetParameterCountException("Incorrect number of parameters specified for TestCase");
+                    }
+                }
+                parms.Arguments = newArgList;
+            }
+#endif
+        }
+
+        /// <summary>
```

Comment:
It's not a good idea to refactor an unrelated class when fixing something different. Remember that the reviewers have to look at every change you make, line by line, to see what it does. There's nothing in GitHub to tell us "He only extracted the method but didn't make changes." OK for this time, but please separate refactoring from features the next time.